File: examples/client-credentials-flow.js
#!/usr/bin/env node

/**
 * OAuth2 Client Credentials Flow Example for Node.js
 * Demonstrates service-to-service authentication using OAuth2
 */

const axios = require('axios');
const qs = require('querystring');

class OAuthClient {
    constructor(clientId, clientSecret, authUrl) {
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.authUrl = authUrl;
        this.httpClient = axios.create({
            timeout: 30000,
            headers: {
                'User-Agent': 'Recipe-Web-App-OAuth-Client/1.0'
            }
        });
        this.token = null;
        this.tokenExpiry = null;
    }

    /**
     * Get access token using client credentials flow
     * @param {string[]} scopes - Requested scopes
     * @returns {Promise<Object>} Token response
     */
    async getAccessToken(scopes = ['read', 'write']) {
        // Check if we have a valid cached token
        if (this.token && new Date() < this.tokenExpiry) {
            return this.token;
        }

        try {
            // Prepare form data
            const data = {
                grant_type: 'client_credentials',
                client_id: this.clientId,
                client_secret: this.clientSecret,
                scope: scopes.join(' ')
            };

            // Make token request
            const response = await this.httpClient.post(
                `${this.authUrl}/api/v1/auth/oauth/token`,
                qs.stringify(data),
                {
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );

            const tokenResponse = response.data;

            // Cache the token with a 60-second buffer before expiry
            this.token = tokenResponse;
            this.tokenExpiry = new Date(Date.now() + (tokenResponse.expires_in - 60) * 1000);

            return tokenResponse;
        } catch (error) {
            if (error.response) {
                throw new Error(`Token request failed: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
            }
            throw new Error(`Token request failed: ${error.message}`);
        }
    }

    /**
     * Make an authenticated HTTP request
     * @param {string} method - HTTP method
     * @param {string} url - Request URL
     * @param {Object} data - Request body (optional)
     * @param {Object} options - Additional axios options
     * @returns {Promise<Object>} Response data
     */
    async makeAuthenticatedRequest(method, url, data = null, options = {}) {
        // Ensure we have a valid access token
        const token = await this.getAccessToken();

        try {
            const config = {
                method,
                url,
                headers: {
                    'Authorization': `${token.token_type} ${token.access_token}`,
                    ...options.headers
                },
                ...options
            };

            if (data) {
                config.data = data;
                config.headers['Content-Type'] = config.headers['Content-Type'] || 'application/json';
            }

            const response = await this.httpClient(config);
            return response.data;
        } catch (error) {
            if (error.response) {
                throw new Error(`Request failed: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
            }
            throw new Error(`Request failed: ${error.message}`);
        }
    }

    /**
     * Introspect a token to validate it
     * @param {string} token - Token to introspect
     * @returns {Promise<Object>} Introspection response
     */
    async introspectToken(token) {
        try {
            const data = {
                token,
                client_id: this.clientId,
                client_secret: this.clientSecret
            };

            const response = await this.httpClient.post(
                `${this.authUrl}/api/v1/auth/oauth/introspect`,
                qs.stringify(data),
                {
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );

            return response.data;
        } catch (error) {
            if (error.response) {
                throw new Error(`Introspection failed: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
            }
            throw new Error(`Introspection failed: ${error.message}`);
        }
    }

    /**
     * Revoke a token
     * @param {string} token - Token to revoke
     * @returns {Promise<void>}
     */
    async revokeToken(token) {
        try {
            const data = {
                token,
                client_id: this.clientId,
                client_secret: this.clientSecret
            };

            await this.httpClient.post(
                `${this.authUrl}/api/v1/auth/oauth/revoke`,
                qs.stringify(data),
                {
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );
        } catch (error) {
            if (error.response) {
                throw new Error(`Revocation failed: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
            }
            throw new Error(`Revocation failed: ${error.message}`);
        }
    }
}

// Example usage
async function main() {
    // Get configuration from environment variables
    const clientId = process.env.CLIENT_ID;
    const clientSecret = process.env.CLIENT_SECRET;
    const authUrl = process.env.AUTH_URL || 'http://localhost:8080';

    if (!clientId || !clientSecret) {
        console.error('❌ CLIENT_ID and CLIENT_SECRET environment variables must be set');
        process.exit(1);
    }

    console.log('OAuth2 Client Credentials Flow Example (Node.js)');
    console.log('==================================================\n');

    try {
        // Create OAuth client
        const client = new OAuthClient(clientId, clientSecret, authUrl);

        // Example 1: Get access token
        console.log('1. Getting access token...');
        const token = await client.getAccessToken(['read', 'write', 'profile']);

        console.log('✅ Access token obtained successfully!');
        console.log(`   Token Type: ${token.token_type}`);
        console.log(`   Expires In: ${token.expires_in} seconds`);
        console.log(`   Scope: ${token.scope || 'N/A'}`);
        console.log(`   Access Token: ${token.access_token.substring(0, 20)}...\n`);

        // Example 2: Introspect the token
        console.log('2. Introspecting token...');
        const introspection = await client.introspectToken(token.access_token);

        console.log('✅ Token introspection successful!');
        console.log(`   Active: ${introspection.active}`);
        console.log(`   Client ID: ${introspection.client_id}`);
        console.log(`   Scope: ${introspection.scope || 'N/A'}`);
        if (introspection.exp) {
            console.log(`   Expires At: ${new Date(introspection.exp * 1000).toISOString()}`);
        }
        console.log();

        // Example 3: Make authenticated request to health endpoint
        console.log('3. Making authenticated request...');
        const healthResponse = await client.makeAuthenticatedRequest(
            'GET',
            `${authUrl}/api/v1/auth/health`
        );

        console.log('✅ Authenticated request successful!');
        console.log(`   Health Status: ${healthResponse.status}`);
        console.log(`   Service: ${healthResponse.service}`);
        console.log(`   Version: ${healthResponse.version || 'N/A'}\n`);

        // Example 4: Demonstrate token caching
        console.log('4. Demonstrating token caching...');
        console.log('   Getting token again (should use cached token)...');

        const start = Date.now();
        const token2 = await client.getAccessToken(['read', 'write']);
        const duration = Date.now() - start;

        console.log('✅ Token retrieved from cache!');
        console.log(`   Duration: ${duration}ms (should be very fast)`);
        console.log(`   Same token: ${token.access_token === token2.access_token}\n`);

        // Example 5: Error handling
        console.log('5. Demonstrating error handling...');
        try {
            await client.makeAuthenticatedRequest('GET', `${authUrl}/api/v1/auth/nonexistent`);
        } catch (error) {
            console.log('✅ Error handling works correctly!');
            console.log(`   Error: ${error.message}\n`);
        }

        console.log('🎉 OAuth2 Client Credentials Flow Example Complete!\n');

        console.log('Usage in your Node.js service:');
        console.log('1. npm install axios');
        console.log('2. Set CLIENT_ID and CLIENT_SECRET environment variables');
        console.log('3. Create an OAuthClient instance');
        console.log('4. Use getAccessToken() to obtain tokens');
        console.log('5. Use makeAuthenticatedRequest() for API calls');
        console.log('6. The client handles token caching and renewal automatically');

    } catch (error) {
        console.error('❌ Example failed:', error.message);
        process.exit(1);
    }
}

// Export the class for use in other modules
module.exports = { OAuthClient };

// Run the example if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

File: examples/client-credentials-flow.go
// Package main demonstrates the OAuth2 client credentials flow
// for service-to-service authentication in Go.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"time"
)

// TokenResponse represents the OAuth2 token response.
type TokenResponse struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	ExpiresIn   int    `json:"expires_in"`
	Scope       string `json:"scope,omitempty"`
}

// OAuthClient handles OAuth2 client credentials flow.
type OAuthClient struct {
	ClientID     string
	ClientSecret string
	AuthURL      string
	HTTPClient   *http.Client
	token        *TokenResponse
	tokenExpiry  time.Time
}

// NewOAuthClient creates a new OAuth2 client.
func NewOAuthClient(clientID, clientSecret, authURL string) *OAuthClient {
	return &OAuthClient{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		AuthURL:      authURL,
		HTTPClient:   &http.Client{Timeout: 30 * time.Second},
	}
}

// GetAccessToken obtains an access token using client credentials flow.
func (c *OAuthClient) GetAccessToken(scopes []string) (*TokenResponse, error) {
	// Check if we have a valid cached token
	if c.token != nil && time.Now().Before(c.tokenExpiry) {
		return c.token, nil
	}

	// Prepare form data for token request
	data := url.Values{}
	data.Set("grant_type", "client_credentials")
	data.Set("client_id", c.ClientID)
	data.Set("client_secret", c.ClientSecret)

	if len(scopes) > 0 {
		scopeStr := ""
		for i, scope := range scopes {
			if i > 0 {
				scopeStr += " "
			}
			scopeStr += scope
		}
		data.Set("scope", scopeStr)
	}

	// Make token request
	resp, err := c.HTTPClient.PostForm(c.AuthURL+"/api/v1/auth/oauth/token", data)
	if err != nil {
		return nil, fmt.Errorf("failed to request token: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var tokenResp TokenResponse
	if err := json.Unmarshal(body, &tokenResp); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}

	// Cache the token with a buffer before expiry
	c.token = &tokenResp
	c.tokenExpiry = time.Now().Add(time.Duration(tokenResp.ExpiresIn-60) * time.Second)

	return &tokenResp, nil
}

// MakeAuthenticatedRequest makes an authenticated HTTP request using the access token.
func (c *OAuthClient) MakeAuthenticatedRequest(method, url string, body interface{}) (*http.Response, error) {
	// Ensure we have a valid access token
	token, err := c.GetAccessToken([]string{"read", "write"})
	if err != nil {
		return nil, fmt.Errorf("failed to get access token: %w", err)
	}

	var reqBody io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewBuffer(jsonBody)
	}

	// Create the request
	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Add authorization header
	req.Header.Set("Authorization", fmt.Sprintf("%s %s", token.TokenType, token.AccessToken))

	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	// Make the request
	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}

	return resp, nil
}

// IntrospectToken validates a token using the introspection endpoint.
func (c *OAuthClient) IntrospectToken(token string) (map[string]interface{}, error) {
	data := url.Values{}
	data.Set("token", token)
	data.Set("client_id", c.ClientID)
	data.Set("client_secret", c.ClientSecret)

	resp, err := c.HTTPClient.PostForm(c.AuthURL+"/api/v1/auth/oauth/introspect", data)
	if err != nil {
		return nil, fmt.Errorf("failed to introspect token: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read introspection response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("introspection failed with status %d: %s", resp.StatusCode, string(body))
	}

	var introspection map[string]interface{}
	if err := json.Unmarshal(body, &introspection); err != nil {
		return nil, fmt.Errorf("failed to parse introspection response: %w", err)
	}

	return introspection, nil
}

func main() {
	// Get configuration from environment variables
	clientID := os.Getenv("CLIENT_ID")
	clientSecret := os.Getenv("CLIENT_SECRET")
	authURL := os.Getenv("AUTH_URL")

	if clientID == "" || clientSecret == "" {
		log.Fatal("CLIENT_ID and CLIENT_SECRET environment variables must be set")
	}

	if authURL == "" {
		authURL = "http://localhost:8080"
	}

	fmt.Printf("OAuth2 Client Credentials Flow Example\n")
	fmt.Printf("======================================\n\n")

	// Create OAuth client
	client := NewOAuthClient(clientID, clientSecret, authURL)

	// Example 1: Get access token
	fmt.Println("1. Getting access token...")
	token, err := client.GetAccessToken([]string{"read", "write", "profile"})
	if err != nil {
		log.Fatalf("Failed to get access token: %v", err)
	}

	fmt.Printf("✅ Access token obtained successfully!\n")
	fmt.Printf("   Token Type: %s\n", token.TokenType)
	fmt.Printf("   Expires In: %d seconds\n", token.ExpiresIn)
	fmt.Printf("   Scope: %s\n", token.Scope)
	fmt.Printf("   Access Token: %s...\n\n", token.AccessToken[:20])

	// Example 2: Introspect the token
	fmt.Println("2. Introspecting token...")
	introspection, err := client.IntrospectToken(token.AccessToken)
	if err != nil {
		log.Fatalf("Failed to introspect token: %v", err)
	}

	fmt.Printf("✅ Token introspection successful!\n")
	fmt.Printf("   Active: %v\n", introspection["active"])
	fmt.Printf("   Client ID: %s\n", introspection["client_id"])
	fmt.Printf("   Scope: %s\n", introspection["scope"])
	if exp, ok := introspection["exp"].(float64); ok {
		fmt.Printf("   Expires At: %s\n", time.Unix(int64(exp), 0))
	}
	fmt.Println()

	// Example 3: Make authenticated request to health endpoint
	fmt.Println("3. Making authenticated request...")
	resp, err := client.MakeAuthenticatedRequest("GET", authURL+"/api/v1/auth/health", nil)
	if err != nil {
		log.Fatalf("Failed to make authenticated request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatalf("Failed to read response: %v", err)
	}

	fmt.Printf("✅ Authenticated request successful!\n")
	fmt.Printf("   Status: %s\n", resp.Status)
	fmt.Printf("   Response: %s\n\n", string(body))

	// Example 4: Demonstrate token caching
	fmt.Println("4. Demonstrating token caching...")
	fmt.Println("   Getting token again (should use cached token)...")

	start := time.Now()
	token2, err := client.GetAccessToken([]string{"read", "write"})
	duration := time.Since(start)

	if err != nil {
		log.Fatalf("Failed to get cached token: %v", err)
	}

	fmt.Printf("✅ Token retrieved from cache!\n")
	fmt.Printf("   Duration: %v (should be very fast)\n", duration)
	fmt.Printf("   Same token: %t\n\n", token.AccessToken == token2.AccessToken)

	fmt.Println("🎉 OAuth2 Client Credentials Flow Example Complete!")
	fmt.Println("\nUsage in your service:")
	fmt.Println("1. Store CLIENT_ID and CLIENT_SECRET as environment variables")
	fmt.Println("2. Create an OAuthClient instance")
	fmt.Println("3. Use GetAccessToken() to obtain tokens")
	fmt.Println("4. Use MakeAuthenticatedRequest() for API calls")
	fmt.Println("5. The client handles token caching and renewal automatically")
}
